### PR TITLE
Refresh token should not be issued if client is not configured with refresh_token grant type#155

### PR DESCRIPTION
**Pull request by lspil**
A refresh token was generated even if the client didn't have the refresh token grant type assigned.
The refresh token is now generated only for a client registered for the refresh token grant type.

Closes gh-155

### CODE REVIEW

**Comment by lspil** [comment]
Hi, This issue seems to be a fault of mine not working often with GitHub. My pull request that solves #155 created this as a new issue. Sorry for this problem.

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationProvider.java
Comment (alek-sys) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationProvider.java:151]: Just my 2 cents here since I was working on the original PR to support refresh token grant.  

I somehow missed this supported authorization grant types. And now I'm not even sure now how these two settings can be separated. It almost feels like `enableRefreshToken` flag can be removed completely and if the grant type is supported then refresh tokens should be issued.

Comment (lspil) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationProvider.java:151]: Yes. I agree. I don't see a reason for that flag. Should I then remove it completely? I think I need some second opinion here before I do that change.

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationProvider.java
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationProvider.java:21]: Please revert this as we don't use wildcard for package imports.

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationProvider.java
Comment (jgrandja) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationProvider.java:151]: I agree as well. Let's remove `enableRefreshToken()` and simply check `registeredClient.getAuthorizationGrantTypes().contains(AuthorizationGrantType.REFRESH_TOKEN)`.

File: oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/client/TestRegisteredClients.java
Comment (jgrandja) [inline oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/client/TestRegisteredClients.java:56]: Let's remove this and reuse an existing `RegisteredClient` template. You can easily adjust the configuration via the returned `Builder`.

**Review by jgrandja — CHANGES_REQUESTED** [review]
Thanks for the PR @lspil !

Just so you are aware of the PR process, after you submit a PR it does create a new type of issue and therefore resulting in 2 issues - the original issue and the PR issue. This is the normal process. After we merge the PR, we will close it and label as a duplicate.

**Comment by jgrandja** [comment]
Thanks for the updates @lspil ! This is now in master.

FYI, I added a polish commit via 8c71e56.

**Pull request closed without merging**
